src/core/device.c

```diff
@@ -811,68 +875,65 @@ static int device_setup_devlink_units(Manager *m, sd_device *dev, char ***ret_re
                 if (device_has_devlink(dev, d->path))
                         continue; /* The devlink was already processed in the above loop. */
 
-                if (device_setup_devlink_unit_one(m, d->path, &assigned) <= 0)
-                        continue;
-
-                if (sd_device_get_syspath(assigned, &s) < 0)
-                        continue;
-
-                if (path_equal(s, syspath))
-                        continue;
-
-                r = strv_extend(&ready_devlinks, d->path);
-                if (r < 0)
-                        return -ENOMEM;
+                (void) device_setup_devlink_unit_one(m, d->path, ready_units, not_ready_units);
         }
 
-        *ret_ready_devlinks = TAKE_PTR(ready_devlinks);
         return 0;
 }
 
-static void device_process_new(Manager *m, sd_device *dev, const char *sysfs) {
-        const char *dn, *alias;
+static int device_setup_units(Manager *m, sd_device *dev, Set **ready_units, Set **not_ready_units) {
+        const char *syspath, *devname = NULL;
         int r;
 
         assert(m);
         assert(dev);
-        assert(sysfs);
+        assert(ready_units);
+        assert(not_ready_units);
 
-        /* Add the main unit named after the sysfs path. If this one fails, don't bother with the rest, as
-         * this one shall be the main device unit the others just follow. (Compare with how
-         * device_following() is implemented, see below, which looks for the sysfs device.) */
-        if (device_setup_unit(m, dev, sysfs, /* main = */ true) < 0)
-                return;
+        r = sd_device_get_syspath(dev, &syspath);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Couldn't get syspath from device, ignoring: %m");
 
-        /* Add an additional unit for the device node */
-        if (sd_device_get_devname(dev, &dn) >= 0)
-                (void) device_setup_unit(m, dev, dn, /* main = */ false);
+        if (device_for_action(dev, SD_DEVICE_REMOVE))
```

Comment:
Can we return early if this is the case? The rest of this function seems to only be adding more units, but are we ever going to add more units on a device remove?

Reply:
No. That is one of the important part of this PR.

Even if a device is removed, a devlink previously points to the device may be still active, as now it may points to another device. Hence, we need to update the corresponding .device unit build from the devlink. That's done at the end of this function.